openedx/core/lib/graph_traversals.py

```diff
@@ -0,0 +1,263 @@
+        try:
+            next_child = current.children.next()
+
+        except StopIteration:
+            # Since there are no children left, visit the node and
+            # remove it from the stack.
+            yield get_result(current.node)
+            visited.add(current.node)
+            stack.pop()
+
+        else:
+            # If so, add the child to the top of the stack.
+            stack.append(_Node(next_child, get_children))
+
+
+def _traverse_generic(
+        start_node,
+        get_parents,
+        get_children,
+        get_result=None,
+        filter_func=None,
+        yield_descendants_of_unyielded=False,
+):
+    """
+    Helper function to avoid duplicating functionality between
+    traverse_depth_first and traverse_topologically.
+
+    If get_parents is None, do a pre-order traversal.
+    Else, do a topological traversal.
+
+    The topological traversal has a worse time complexity than
+    pre-order does, as it needs to check whether each node's
+    parents have been visited.
+
+    Arguments:
+        See description in traverse_topologically.
+    """
+
+    # If get_result or filter_func aren't provided, make them no-ops.
+    get_result = get_result or (lambda node_: node_)
+    filter_func = filter_func or (lambda __: True)
+
+    # Use deque for the stack, which is O(1) for pop and append.
+    stack = deque([start_node])
+
+    # Keep track of which nodes have been visited and whether they
+    # were in fact yielded.
+    yield_results = {}  # dict(node:boolean)
+
+    # While there are more nodes on the stack...
+    while stack:
+
+        # Take a node off the top of the stack.
+        current_node = stack.pop()
+
+        # If we're doing a topological traversal, then make sure all
+        # the node's parents have been visited. If they haven't,
+        # then skip the node for now; we'll encounter it again later
+        # through another one of its parents.
+        if get_parents and current_node != start_node:
+            parents = get_parents(current_node)
+
+            # If all of the parents have not yet been visited, continue.
+            if not all(parent in yield_results for parent in parents):
+                continue
+
+            # If none of the parents have yielded, continue, unless
+            # specified otherwise (via yield_descendants_of_unyielded).
+            elif not yield_descendants_of_unyielded and not any(yield_results[parent] for parent in parents):
+                continue
+
+        # If the current node has already been visited, continue.
+        if current_node not in yield_results:
+
+            if get_parents:
+                # For a topological sort, it's important that we visit the
+                # children even if the parent isn't yielded in case a
+                # child has multiple parents and this is its last parent.
+                unvisited_children = list(get_children(current_node))
```

Comment:
For a topological sort, none of the children should have yield_results at this point since this parent isn't yet visited.
This isn't clear in the comment just preceding this line.  The comment here is explaining why we stack the children *before* checking the yield_result of the parent.  I can update the comments to make this clear.

Overall, this implementation would be much simpler and better optimized for pre-order traversal once we don't have to worry about DAGs.